### PR TITLE
upload profile images to s3, update user in backend

### PR DESCRIPTION
**Pull request by mrota97**

### CODE REVIEW

File: backend/src/routes/user.ts [hunk 1]
```diff
@@ -3,10 +3,15 @@ import { Request, Response, Router } from 'express';
...
 import { IUser, User } from '../models/user';
+import { upload } from '../s3';
 import { handleErrors } from '../util';
 
 const saltLength = 11;
 
+interface IMulterFile {
+    location: string;
+}
+
```
Comment (benwaffle) [inline backend/src/routes/user.ts]: you shouldn't create this interface, you should import it from @types/multer-s3.

File: backend/src/server.ts [hunk 2]
```diff
@@ -16,6 +16,7 @@ mongoose.connect('mongodb://localhost/mothballs', { useNewUrlParser: true })
 
 app.use(morgan('dev'));
 app.use(cors());
+app.use(bodyParser.urlencoded({ extended: true }));
```
Comment (benwaffle) [inline backend/src/server.ts]: did we end up needing this?

File: client/src/actions/auth.js [hunk 3]
```diff
@@ -44,7 +44,15 @@ export function registerUser(username, password) {
...
-      const res = await axios.patch(`http://localhost:4000/api/user/${userId}`, {data})
+      console.log(data);
+      const res = await axios.patch(`http://localhost:4000/api/user/${userId}`, data);
+      // const res = await fetch(
+      //   `http://localhost:4000/api/user/${userId}`,
+      //   {
+      //     method: 'PATCH',
+      //     body: data
+      //   }
+      // );
```
Comment (benwaffle) [inline client/src/actions/auth.js]: remove commented code

File: client/src/components/Navbar.jsx [hunk 4]
```diff
@@ -51,7 +51,7 @@ class Navbar extends React.Component {
             {user._id
             ? <>
                 <a className="navbar-item" onClick={this.logout}>Logout {user.username}</a>
-              <Link className="navbar-item" to="/account/profile"><img src={user.profile_image} alt="profile" /></Link>
+              <NavLink className="navbar-item" to="/account/profile"><img src={user.profile_image} alt="profile" /></NavLink>
```
Comment (benwaffle) [inline client/src/components/Navbar.jsx]: indentation

File: client/src/components/pages/Account/Profile.jsx [hunk 5]
```diff
@@ -3,19 +3,25 @@ import { connect } from 'react-redux';
...
-    username: this.props.user.username,
-    profile_image: this.props.user.profile_image,
-    bio: this.props.user.bio,
-    image_upload: null,
+    username: '',
+    profile_image: null,
+    bio: '',
     image_name: '',    
     error: null
   };
```
Comment (benwaffle) [inline client/src/components/pages/Account/Profile.jsx]: just do `state = initialState`

File: client/src/components/pages/Account/Profile.jsx [hunk 6]
```diff
@@ -30,15 +36,16 @@ class Profile extends React.Component {
     event.preventDefault();
 
     const { user } = this.props;
+    const { username, profile_image, bio, image_name } = this.state;
+
+    let form_data = new FormData();
```
Comment (benwaffle) [inline client/src/components/pages/Account/Profile.jsx]: const

File: client/src/components/pages/Account/Profile.jsx [hunk 7]
```diff
@@ -47,89 +54,91 @@ class Profile extends React.Component {
...
-                    className="textarea"
-                    onChange={this.handleChange}
-                    placeholder={user.bio}
-                    value={update.bio} />
+            <div className="column">
+              <form onSubmit={this.handleSubmit} encType="multipart/form-data">
+                <h1>Update profile</h1>
+
+                <div className="field">
+                  <label htmlFor="" className="label">username</label>
```
Comment (benwaffle) [inline client/src/components/pages/Account/Profile.jsx]: remove `htmlFor=""`

File: client/src/components/pages/Account/Profile.jsx [hunk 8]
```diff
@@ -47,89 +54,91 @@ class Profile extends React.Component {
...
+                          <span className="file-label">
+                            Choose a file…
+                          </span>
+                        </span>
+                        <span className="file-name">
+                          {this.state.profile_image
+                          ? this.state.image_name
+                          : <>
+                              no image uploaded
+                            </>}
```
Comment (benwaffle) [inline client/src/components/pages/Account/Profile.jsx]: use a string `'no image uploaded'`, not a fragment

File: client/src/components/pages/Account/Profile.jsx [hunk 9]
```diff
@@ -47,89 +54,91 @@ class Profile extends React.Component {
...
+                        </span>
+                    </label>
+                  </div>  
+                </div>
+
+                  <div className="field">
+                    <div className="control">
+                      <button className="button" type="submit" value="submit">Update</button>
+                    </div>
+                  </div>
```
Comment (benwaffle) [inline client/src/components/pages/Account/Profile.jsx]: indentation

File: client/src/reducers/exerciseEvents.js [hunk 10]
```diff
@@ -6,7 +6,7 @@ export default function exerciseEventsReducer(state = initialState.exerciseEvent
     case EXERCISE_EVENT_CREATED:
       return [...state, action.data]
     case USER_LOGIN_SUCCESS:
-      return action.data.exerciseEvents
+      return action.data.exerciseEvents || null
```
Comment (benwaffle) [inline client/src/reducers/exerciseEvents.js]: ~i think you need to return exerciseEvents in the update route as well~

do `return action.data.exerciseEvents || state` instead

File: backend/src/routes/user.ts (hunk 1 above)
Comment (mrota97) [reply backend/src/routes/user.ts]: done

File: backend/src/server.ts (hunk 2 above)
Comment (mrota97) [reply backend/src/server.ts]: no

**Comment by cbouwense** [comment]
bepis

**Pull request merged**